### PR TITLE
Initial development

### PR DESCRIPTION
**Pull request by jeffswenson**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+* gulp-cli (can be installed through npm)
+* tsd (can be installed through npm)
+
+# Install instructions
+In the future this will likely be wrapped up in an npm package.  Until then you will need to build and install it manually.
+
+1. Clone the repository onto your machine.
+2. Change into the directory.
+3. Run `npm install`
+4. Change into the TSD file and run `tsd install`
```
Comment (tlein) [inline README.md]: This should be changed to "TSD folder"

File: README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+5. Add the following to your bashrc:
+5. Add release/lib to your NODE_PATH environment variable.
+6. Add release/bin to your PATH environment variable.
```
Comment (tlein) [inline README.md]: When do I run gulp?

File: Src/Raft/CLIActions.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        });
+    });
+}
+
+/**
+ * Create a new project in the current directory
+ * @return {Promise<any>} A promise that resolves when the new project is ready.
+ */
+export function create() : Promise<any> {
+    raftlog("Project", "Creating a new project");
```
Comment (tlein) [inline Src/Raft/CLIActions.ts:47]: If you have a space in your directory at any point this will fail.

File: Src/Raft/RaftFile.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+}
+
+/**
+ * Create a Repostitory object given a descriptor from a raftfile.
+ * @param  {RaftFile.RepositoryDescriptor} repoDescriptor The descriptor loaded from a raftfile.
+ * @return {VCS.Repository}                               An object that can be used to interact with the repository.
+ */
+export function createRepository(repoDescriptor : RepositoryDescriptor) : VCS.Repository {
+    console.log("Repo Branch: " + repoDescriptor.branch);
```
Comment (tlein) [inline Src/Raft/RaftFile.ts]: When pulling from master this said "undefined"

File: Doc/DESIGN [hunk 5]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (tlein) [inline Doc/DESIGN]: Delete this file

File: Src/CLI/RaftCLI.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env node
+var Raft = require("Raft");
+
+var yargs = require('yargs')
+    .usage('Usage: raft <command> [options]')
+    .command('build', 'Build the raft project')
+    .command('create', 'Create a new project')
+    .boolean(['create']);
```
Comment (tlein) [inline Src/CLI/RaftCLI.js]: Should create be a boolean flag as well?

File: Src/Raft/CLIActions.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import _ = require('underscore');
+
+import BuildConfig = require('./BuildConfig');
+import Dependency = require('./Dependency');
+import Path = require('./Path');
+import Project = require('./Project');
+import Promise = require('bluebird');
```
Comment (tlein) [inline Src/Raft/CLIActions.ts]: Move bluebird above with underscore

File: Src/Raft/CLIActions.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Build the raft project the user is currently in.
+ * @param  {object} options Can be used to specify the parameters for the build configuration.
+ * @return {Promise}        A promise that resolves once the build is finished.
+ */
+export function build(options : {platform? : string, architecture? : string} = {}) : Promise<any> {
+    var buildSettings : BuildConfig.Build = {
+        isDeploy : false,
+        platform : BuildConfig.Platform.Host,
+        architecture : "host"
```
Comment (tlein) [inline Src/Raft/CLIActions.ts]: Make architecture an enum like the platform

File: Src/Raft/CLIActions.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Create a new project in the current directory
+ * @return {Promise<any>} A promise that resolves when the new project is ready.
+ */
+export function create() : Promise<any> {
+    raftlog("Project", "Creating a new project");
+    var repo = new VCS.GitRepository("https://github.com/tlein/AnconaTemplateGame");
+    var destinationDir = Path.cwd();
+
+    var context = {
```
Comment (tlein) [inline Src/Raft/CLIActions.ts:67]: Prompt the user for these values

File: Src/Raft/CMake.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
+import _ = require('underscore');
+
+import BuildConfig = require('./BuildConfig');
+import Path = require('./Path');
+import System = require('./System');
+
+module CMake {
+
+    export const FRAMEWORK_PREFIX = "framework";
```
Comment (tlein) [inline Src/Raft/CMake.ts]: This isn't used as a prefix, rename it to FRAMEWORK_DIRECTORY.

File: Src/Raft/CMake.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    const RAFT_LIB_DIR = "RAFT_LIB_DIR";
+    const RAFT_FRAMEWORK_DIR = "RAFT_FRAMEWORK_DIR";
+    const RAFT_IS_DESKTOP = "RAFT_IS_DESKTOP";
+    const RAFT_IS_ANDROID = "RAFT_IS_ANDROID";
+
+    /**
+     * Configure a cmake project.
+     * @param  {Path}         srcPath   Root of the cmake project.
+     * @param  {Path}         buildPath Directory the configuration was stored in.
+     * @param  {any}          options   Set of Key Values that will be passed to CMake as configuration options.
```
Comment (tlein) [inline Src/Raft/CMake.ts]: This is not an `any` object but rather a `CMakeOptions` object

File: Src/Raft/CMake.ts [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            }
+            return result;
+        }
+
+        /**
+         * Return an array of arguments that can be passed to a cmake command.
+         * @return {string} Options set as CMake flags.
+         */
+        toArray() : string [] {
+            return _.chain(_.pairs(this.options))
```
Comment (tlein) [inline Src/Raft/CMake.ts]: Not sure if this chain is necessary

File: Src/Raft/Path.ts [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        return path != null && path.path === this.path;
+    }
+
+    /**
+     * Create the directory described by the path.
+     * @return {Promise<boolean>} Promise that resolves to true if the path was
+     *                            created and false if it already existed.
+     */
+    createDirectory() : Promise<boolean> {
+        return Promise.fromNode((callback) => {
```
Comment (tlein) [inline Src/Raft/Path.ts]: Should use `fromCallback` instead of `fromNode`

File: Src/Raft/Project.ts [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    /**
+    *  Will walk up the file tree until the root directory of a project is found.
+    *  @param path The path to start the search from
+    *  @return Returns the project if it exists or null if it does not.
+    */
+    static find(path : Path) : Promise<Project> {
+        var paths = [path];
+        var raftDir = new Path("Raft");
+
+        while (!_.last(paths).isRoot()) {
```
Comment (tlein) [inline Src/Raft/Project.ts]: This block could be simplified according to you

File: Src/Raft/Project.ts [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    /**
+     * Load a project from the raftfile.
+     * @return {Promise<Project>} A promise that resolves to a loaded project.
+     */
+    load() : Promise<Project> {
+        return this.root.append(Project.RAFT_FILE).read()
+        .then((data) => {
+            raftlog("Project Data", data);
+            this.raftfile = JSON.parse(data);
+            raftlog("Project Data", JSON.stringify(this.raftfile));
```
Comment (tlein) [inline Src/Raft/Project.ts]: Remove these logs

File: Src/Raft/Project.ts [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+            return this;
+        });
+    }
+
+    /**
+     * Get the dependency descriptor's contained in the project's raftfile.
+     * @return {RaftFile.DependencyDescriptor} [description]
```
Comment (tlein) [inline Src/Raft/Project.ts]: Incomplete header doc

File: Doc/DESIGN (hunk 5 above)
Comment (jeffswenson) [inline Doc/DESIGN]: Done

File: README.md (hunk 1 above)
Comment (jeffswenson) [inline README.md]: done

File: README.md (hunk 2 above)
Comment (jeffswenson) [inline README.md]: done

File: Src/CLI/RaftCLI.js (hunk 6 above)
Comment (jeffswenson) [inline Src/CLI/RaftCLI.js]: done

File: Src/Raft/Project.ts [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+     * @return {[type]}                  Path describing where the framework headers should be installed.
+     */
+    dirForDependencyFramework(build : BuildConfig.Build) {
+        return this.dirForDependencyInstall(build).append(Project.DEPENDENCY_FRAMEWORK_DIR);
+    }
+
+    /**
+     * Get the directory the project should be built in.
+     * @param  {BuildConfig.Build} build The current build configuration.
+     * @return {Path}                  [description]
```
Comment (tlein) [inline Src/Raft/Project.ts]: Incomplete header doc

File: Src/Raft/CLIActions.ts (hunk 8 above)
Comment (jeffswenson) [inline Src/Raft/CLIActions.ts]: done

File: Src/Raft/System.ts [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Execute a command in a child process.
+ * @param  {string}                 command  Shell command to execute.
+ * @param  {string []}              args     Array of arguments that will be passed to the command. Note: The args argument is not optional in order to remind people not to pass arguments by concatenating them to the command string.
+ * @param  {ExecuteOptions}         options  (Optional) @see ExecuteOptions
+ * @return {Promise<ProcessOutput>}          Promise that resolves to the process's output.
+ */
+export function execute(command : string, args : string [], options? : ExecuteOptions) : Promise<ProcessOutput> {
+    options = options || {};
+    var start = Promise.resolve();
```
Comment (tlein) [inline Src/Raft/System.ts]: I think this is dumb, but I don't know enough about promises to be sure.

File: package.json [hunk 19]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "Raft",
+  "version": "0.0.0",
```
Comment (tlein) [inline package.json]: Update to 0.0.1

File: Src/Raft/CLIActions.ts (hunk 7 above)
Comment (jeffswenson) [inline Src/Raft/CLIActions.ts]: done

File: Src/Raft/CLIActions.ts (hunk 3 above)
Comment (jeffswenson) [inline Src/Raft/CLIActions.ts:47]: done

File: Src/Raft/CLIActions.ts (hunk 9 above)
Comment (jeffswenson) [inline Src/Raft/CLIActions.ts:67]: done

File: Src/Raft/CMake.ts (hunk 10 above)
Comment (jeffswenson) [inline Src/Raft/CMake.ts]: done

File: Src/Raft/CMake.ts (hunk 11 above)
Comment (jeffswenson) [inline Src/Raft/CMake.ts]: done

File: Src/Raft/CMake.ts (hunk 12 above)
Comment (jeffswenson) [inline Src/Raft/CMake.ts]: done

File: Src/Raft/Path.ts (hunk 13 above)
Comment (jeffswenson) [inline Src/Raft/Path.ts]: done

File: Src/Raft/Project.ts (hunk 14 above)
Comment (jeffswenson) [inline Src/Raft/Project.ts]: done

File: Src/Raft/Project.ts (hunk 15 above)
Comment (jeffswenson) [inline Src/Raft/Project.ts]: done

File: Src/Raft/Project.ts (hunk 16 above)
Comment (jeffswenson) [inline Src/Raft/Project.ts]: done

File: Src/Raft/Project.ts (hunk 17 above)
Comment (jeffswenson) [inline Src/Raft/Project.ts]: done

File: Src/Raft/RaftFile.ts (hunk 4 above)
Comment (jeffswenson) [inline Src/Raft/RaftFile.ts]: done

File: Src/Raft/System.ts (hunk 18 above)
Comment (jeffswenson) [inline Src/Raft/System.ts]: improved the code

File: package.json (hunk 19 above)
Comment (jeffswenson) [inline package.json]: done

File: Src/Raft/CLIActions.ts (hunk 3 above)
Comment (tlein) [inline Src/Raft/CLIActions.ts:47]: confirmed

File: Src/Raft/RaftFile.ts (hunk 4 above)
Comment (tlein) [inline Src/Raft/RaftFile.ts]: confirmed

File: Doc/DESIGN (hunk 5 above)
Comment (tlein) [inline Doc/DESIGN]: confirmed

File: Src/CLI/RaftCLI.js (hunk 6 above)
Comment (tlein) [inline Src/CLI/RaftCLI.js]: confirmed

File: Src/Raft/CLIActions.ts (hunk 7 above)
Comment (tlein) [inline Src/Raft/CLIActions.ts]: confirmed

File: Src/Raft/CLIActions.ts (hunk 8 above)
Comment (tlein) [inline Src/Raft/CLIActions.ts]: confirmed

File: Src/Raft/CLIActions.ts (hunk 9 above)
Comment (tlein) [inline Src/Raft/CLIActions.ts:67]: confirmed

File: Src/Raft/CMake.ts (hunk 10 above)
Comment (tlein) [inline Src/Raft/CMake.ts]: confirmed

File: Src/Raft/CMake.ts (hunk 11 above)
Comment (tlein) [inline Src/Raft/CMake.ts]: confirmed

File: Src/Raft/CMake.ts (hunk 12 above)
Comment (tlein) [inline Src/Raft/CMake.ts]: confirmed

File: Src/Raft/Path.ts (hunk 13 above)
Comment (tlein) [inline Src/Raft/Path.ts]: confirmed

File: Src/Raft/Project.ts (hunk 14 above)
Comment (tlein) [inline Src/Raft/Project.ts]: confirmed

File: Src/Raft/Project.ts (hunk 15 above)
Comment (tlein) [inline Src/Raft/Project.ts]: confirmed

File: Src/Raft/Project.ts (hunk 16 above)
Comment (tlein) [inline Src/Raft/Project.ts]: confirmed

File: Src/Raft/Project.ts (hunk 17 above)
Comment (tlein) [inline Src/Raft/Project.ts]: confirmed

File: Src/Raft/System.ts (hunk 18 above)
Comment (tlein) [inline Src/Raft/System.ts]: confirmed

File: package.json (hunk 19 above)
Comment (tlein) [inline package.json]: confirmed

**Pull request merged**